sample_files/main.py

```diff
@@ -1,145 +1,80 @@
-      try:
-          # Initialize Selenium WebDriver
-          driver = webdriver.Chrome()  
-          driver.get(url)  
-
-          # Check if the elements interacted with by the chain are present on the webpage
-          for result in chain_results:
-              element = driver.find_element(By.XPATH, f"//*[contains(text(), '{result}')]")
-              if not element:
-                  logger.error(f"Element {result} not found on the webpage")
-                  return False
-
-          driver.quit()
-          return True
-
-      except WebDriverException as e:
-          logger.error(f"Selenium WebDriver error: {e}")
-          return False
-      
-
-  
-def execute_chain(urls:list, max_attempts=5, attempt=1):
-    """Execute a chain process, retries if it fails, up to max_attempts."""
+        try:
+            driver = get_webdriver()
+            driver.get(url)
+            for result in chain_results:
+                element = driver.find_element(By.XPATH, f"//*[contains(text(), '{result}')]")
+                if not element:
+                    logger.error(f"Element {result} not found on the webpage {url}")
+                    driver.quit()
+                    return False
+            driver.quit()
+            return True
+        except WebDriverException as e:
+            logger.error(f"Selenium WebDriver error: {e}")
+            return False
+
+# Function to execute the chain process
+def execute_chain(urls, max_attempts=5, attempt=1):
     if attempt > max_attempts:
         logger.error("Max attempts reached. Exiting.")
         return
 
     try:
-        # Checking web application status
         response = requests.get(urls[0])
         response.raise_for_status()
 
-        # Invoking chain results
-        logger.info("Chain execution has started . . . ")
-
-        # Capture relevant information during chain execution
-        chain_results = chain.invoke({"urls": read_urls(urls)})  # chain invoke
+        logger.info("Starting chain execution...")
+        chain_results = chain.invoke({"urls": read_urls(urls)})
         time.sleep(2)
+        logger.info("Chain execution successful.")
 
-        logger.info("Chain has successfully executed.")
-
-        # Check if Selenium process executed properly
-        if not check_chain_interactions(chain_results,urls):
+        if not check_chain_interactions(chain_results, urls):
             raise RuntimeError("Selenium process did not execute properly")
 
-        return 
+        return
 
     except (RequestException, ValueError, RuntimeError) as e:
         logger.error(f"Error occurred: {e}")
-        print(f"Error occurred: {e}")
+        logging.error(f"Error occurred: {e}")
 
-    # If selenium didnt execute properly or validation failed, retry
     logger.info(f"Retrying... Attempt {attempt}/{max_attempts}")
-    time.sleep(3)  # wait for some time before re-executing
+    time.sleep(3)
     execute_chain(urls, max_attempts, attempt + 1)
 
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

In the `execute_chain` function, there is a redundant logging statement on line 73. It seems to be a duplicate of the previous line and should be removed.

```diff
- logging.error(f"Error occurred: {e}")
```



<!-- This is an auto-generated comment by OSS CodeRabbit -->